# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Adrian Vega
### GitHub Repository: https://github.com/adriacv17/json-sentime

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

## Question 1

The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [4]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package            Version
------------------ ------------
annotated-types    0.7.0
asttokens          2.4.1
blis               1.0.1
catalogue          2.0.10
certifi            2024.8.30
charset-normalizer 3.4.0
click              8.1.7
cloudpathlib       0.20.0
colorama           0.4.6
comm               0.2.2
confection         0.1.5
cymem              2.0.8
debugpy            1.8.8
decorator          5.1.1
en_core_web_sm     3.8.0
executing          2.1.0
idna               3.10
ipykernel          6.29.5
ipython            8.29.0
jedi               0.19.2
Jinja2             3.1.4
joblib             1.4.2
jupyter_client     8.6.3
jupyter_core       5.7.2
langcodes          3.4.1
language_data      1.2.0
marisa-trie        1.2.1
markdown-it-py     3.0.0
MarkupSafe         3.0.2
matplotlib-inline  0.1.7
mdurl              0.1.2
murmurhash         1.0.10
nest-asyncio       1.6.0
nltk               3.9.1
numpy              2.0.2
packaging          24.2
parso     

In [ ]:
# import libraries

import json
import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Retrieve lyrics data from API
result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)

# Write result as a JSON file named lyrics_object.json
with open('lyrics_object.json', 'w') as json_file:  # 'w' writes lyrics from result by dumping into json file
    json.dump(result, json_file, indent=4)

# Load the JSON data from the file and read using 'r'
with open('lyrics_object.json', 'r') as json_file:
    lyrics_data = json.load(json_file) 

print(lyrics_data) # verify the lyrics are in variable


{'lyrics': "I'm your only friend \nI'm not your only friend \nBut I'm a little glowing friend \nBut really I'm not actually your friend \nBut I am \n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI have a secret to tell \n\nFrom my electrical well \n\nIt's a simple message and I'm leaving out the whistles and bells \n\nSo the room must listen to me \n\nFilibuster vigilantly \n\nMy name is blue canary one note* spelled l-i-t-e \n\nMy story's infinite \n\nLike the Longines Symphonette it doesn't rest \n\n\n\nBlue canary in the outlet by the light switch \n\nWho watches over you \n\nMake a little birdhouse in your soul \n\nNot to put too fine a point on it \n\nSay I'm the only bee in your bonnet \n\nMake a little birdhouse in your soul \n\n\n\nI'm your only friend \n\nI'm not your only friend \n\

## Question 2

Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [ ]:

# import libraries

import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load the JSON data from the file and read using 'r'
with open('lyrics_object.json', 'r') as json_file:
    lyrics_data = json.load(json_file)

# Load spacy model and create spacy pipeline
nlp = spacy.load('en_core_web_sm') 
nlp.add_pipe('spacytextblob')

# Get the lyrics from the dictionary using .get function
lyrics = lyrics_data.get('lyrics')

# Process the lyrics
model_doc = nlp(lyrics)

# Perform sentiment analysis and get the polarity score
sentiment = model_doc._.blob.polarity

# Print the lyrics and the polarity score
print('Lyrics:\n') # add \n to create a new line after Lyrics: prints
print(lyrics) # print lyrics from lyrics variable
print()
print()
print()
print('\nPolarity score:', sentiment) # use \n before print Polarity score to create new line before print statement

# Create a conditional statement to create variable comment giving sentiment comment based on score, closer to 1 is positive, closer to -1 is negative
if sentiment > 0:
    comment = 'positive'
elif sentiment < 0:
    comment = 'negative'
else:
    comment = 'neutral'

print() # add an empty line
print('The lyrics have a', comment, 'connotation.') # create print statement for sentiment connotation 



Lyrics:

I'm your only friend 
I'm not your only friend 
But I'm a little glowing friend 
But really I'm not actually your friend 
But I am 


Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I have a secret to tell 

From my electrical well 

It's a simple message and I'm leaving out the whistles and bells 

So the room must listen to me 

Filibuster vigilantly 

My name is blue canary one note* spelled l-i-t-e 

My story's infinite 

Like the Longines Symphonette it doesn't rest 



Blue canary in the outlet by the light switch 

Who watches over you 

Make a little birdhouse in your soul 

Not to put too fine a point on it 

Say I'm the only bee in your bonnet 

Make a little birdhouse in your soul 



I'm your only friend 

I'm not your only friend 

But I'm a little glowing friend 

But really I'm not actual

## Question 3

Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [1]:
# import libraries

import json
import requests

def get_lyrics(artist, song, file):
    # Create the API URL based on artist and song
    url = f'https://api.lyrics.ovh/v1/{artist}/{song}' 
    
    # Make a request to the API
    response = requests.get(url)
        
    # If the request was successful with code being 200 parse the JSON response
    if response.status_code == 200:
        result = response.json()              
        # Write the lyrics to the file
        with open(file, 'w') as json_file:
            json.dump(result, json_file, indent=4)
                
        print(f"Lyrics for '{song}' by {artist} saved to {file}.")
    # If request not successful give error comment   
    else:
        print(f"Could not get lyrics for '{song}' by {artist} due to error code: {response.status_code}")


# Test the function with four songs saved to files

# Song 1 "Speed of Sound" by Coldplay
get_lyrics('Coldplay', 'Speed of Sound', 'speed_of_sound.json')

# Song 2 "Midnight City" by M83
get_lyrics('M83', 'Midnight City', 'midnight_city.json')

# Song 3 "Radioactive" by Imagine Dragones
get_lyrics('Imagine Dragons', 'Radioactive', 'radioactive.json')

# Song 4 "In the End" by Linkin Park
get_lyrics('Linkin Park', 'In the End', 'in_the_end.json')


Lyrics for 'Speed of Sound' by Coldplay saved to speed_of_sound.json.
Lyrics for 'Midnight City' by M83 saved to midnight_city.json.
Lyrics for 'Radioactive' by Imagine Dragons saved to radioactive.json.
Lyrics for 'In the End' by Linkin Park saved to in_the_end.json.


## Question 4

Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [3]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load spacy model and SpacyTextBlob
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("spacytextblob")

def sentiment_analysis(file, song):
    # Load the lyrics from JSON
    with open(file, 'r') as json_file:
        lyrics_data = json.load(json_file)
    
    # Get the lyrics
    lyrics = lyrics_data.get('lyrics', '')
    
    # Sentiment analysis
    doc = nlp(lyrics)
    sentiment = doc._.blob.polarity
    
    # Print song name and its polarity score as sentiment
    print(f"Polarity Score for lyrics in '{song}': {sentiment}")
    return sentiment

# Test the function with 3 JSON files
sentiment_analysis("speed_of_sound.json", "Speed of Sound")
sentiment_analysis("midnight_city.json", "Midnight City")
sentiment_analysis("radioactive.json", "Radioactive")
sentiment_analysis("in_the_end.json", "In the End")

Polarity Score for lyrics in 'Speed of Sound': 0.24047619047619043
Polarity Score for lyrics in 'Midnight City': 0.08660714285714285
Polarity Score for lyrics in 'Radioactive': 0.28161838161838165
Polarity Score for lyrics in 'In the End': -0.019340924775707397


-0.019340924775707397

## Question 4B

Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

## Answer 4B

Song Speed of Sound: Spacy gave it a sentiment of 0.24 which is slightly positive, but more neutral. I believe this song has more of a negative connotation. I believe the sentiment differs because Spacy can't hear the undertones of the song and is taking certain words that may need a deeper understanding than ranking the words on polarity.
<br>
<br>
Song Midnight City: Spacy gave it a sentiment of 0.08 which is neutral to slightly positive. I agree completelty with Spacy as the song does reflect very neutral based on its tone and words.
<br>
<br>
Song Radioactive: Spacy gave it a sentiment of 0.28 which is more positive than neutral. I agree with the Spacy as the song may be in an apocolyptic world it has words of empowerment and awakening.
<br>
<br>
Song In the End: Spacy gave it a sentiment of -0.019 which is slightly negative. I believe it has more negative sentiment than it is credited for because of all the words of struggle and hopelessness. I believe certain words can add to the positive reflection of the song while it is more of a negative connotation.
